Adds UA Quirk for open.spotify.com on iPad Mini

karlcow opened the pull request:
#### 5ce3668f4088a5297a2d6cb9135907456036fa3f
<pre>
Adds UA Quirk for open.spotify.com on iPad Mini
<a href="https://bugs.webkit.org/show_bug.cgi?id=297710">https://bugs.webkit.org/show_bug.cgi?id=297710</a>
<a href="https://rdar.apple.com/158172517">rdar://158172517</a>

Reviewed by Brandon Stewart and Brent Fulgham.

iPad Mini users are being blocked because of the UA string. This patch
makes it possible for iPad Mini users to receive the iPhone version of
the Spotify website, which would be better than the desktop version
given the form factor of the iPad Mini.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsIPhoneUserAgent):

Canonical link: <a href="https://commits.webkit.org/299301@main">https://commits.webkit.org/299301@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eea0a3fd93afe07bc255dd32fd90d0e134fc3077

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118550 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38231 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28882 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124724 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/70609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4060abc0-4aca-4801-8424-ba07bb7b557c) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120428 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38927 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46813 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89983 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/70609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/14e274c3-e9ae-4579-ae2f-ccee039457d4) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121503 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31004 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/106291 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70487 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c2f9a08a-541a-4566-961c-23b94703a04c) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/30060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24402 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68381 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100443 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24593 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/127788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45457 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/34294 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/127788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45821 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102512 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/127788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43838 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21831 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/41955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18890 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45327 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51005 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/44790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/48137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46477 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1459,13 +1460,16 @@ bool Quirks::needsIPadMiniUserAgent(const URL& url)
 }
 
-bool Quirks::needsIPhoneUserAgent(const URL& url)
+bool Quirks::needsIPhoneUserAgent(API::WebsitePolicies& policies, const URL& url)
```
Where is this used?

brentfulgham replied:
Yes, Brandon is right. It doesn't seem like the `policies` object (which is a WebKit-level concept, so a layering violation to user here) is needed. If you remove it, I think this patch will build without problem.

karlcow replied:
thanks.

stwrt commented on Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
```diff
@@ -1470,8 +1470,11 @@ static RecommendDesktopClassBrowsingForRequest desktopClassBrowsingRecommendedFo
 
-    if (Quirks::needsIPhoneUserAgent(request.url()))
-        return RecommendDesktopClassBrowsingForRequest::No;
+    if (Quirks::needsIPhoneUserAgent(API::WebsitePolicies& policies, request.url())) {
```
Is this needed here `API::WebsitePolicies&`?

karlcow replied:
nope. Thanks

brentfulgham commented on Source/WebCore/page/Quirks.h:
```diff
@@ -110,7 +111,7 @@ class Quirks {
     WEBCORE_EXPORT String storageAccessUserAgentStringQuirkForDomain(const URL&);
     WEBCORE_EXPORT static bool needsIPadMiniUserAgent(const URL&);
-    WEBCORE_EXPORT static bool needsIPhoneUserAgent(const URL&);
+    WEBCORE_EXPORT static bool needsIPhoneUserAgent(API::WebsitePolicies&, const URL&);
```
It doesn't seem like you need `policies` here at all.

karlcow replied:
ha yes.

brentfulgham commented on Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
```diff
@@ -1470,8 +1470,11 @@ static RecommendDesktopClassBrowsingForRequest desktopClassBrowsingRecommendedFo
-        return RecommendDesktopClassBrowsingForRequest::No;
+    if (Quirks::needsIPhoneUserAgent(API::WebsitePolicies& policies, request.url())) {
+        policies.setCustomUserAgent(makeStringByReplacingAll(standardUserAgentWithApplicationName(m_applicationNameForUserAgent), "iPad"_s, "iPhone"_s));
+        policies.setCustomNavigatorPlatform("iPhone"_s);
```
You use the `policies` object here, which is fine since it is in the UIProcess (at the WebKit level). No need to pass it to the `WebCore::Quirks` layer as far as I can tell.

karlcow replied:
yes, I overthought it. 
```
const API::WebsitePolicies& policies = m_configuration->defaultWebsitePolicies();
```

brentfulgham reviewed (changes requested):
Just remove the unneeded `policies` stuff in `WebCore::Quirks` and I think this is fine. If you need the `Quirks` code to know something like whether the policy is asking for a mobile site or something, just pass that as a flag.

You aren't allowed to use WebKit concepts down in WebCore, but you can pass state through flags if you need to.

stwrt commented on Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
```diff
@@ -1470,8 +1470,12 @@ static RecommendDesktopClassBrowsingForRequest desktopClassBrowsingRecommendedFo
+        const API::WebsitePolicies& policies = m_configuration->defaultWebsitePolicies();
+        policies.setCustomUserAgent(makeStringByReplacingAll(standardUserAgentWithApplicationName(m_applicationNameForUserAgent), "iPad"_s, "iPhone"_s));
+        policies.setCustomNavigatorPlatform("iPhone"_s);
+        return WebContentMode::Mobile;
```
Why are we returning a WebContentMode here?

The function returns a RecommendDesktopClassBrowsingForRequest.

brentfulgham replied:
Yes: You should change this to:
`return RecommendDesktopClassBrowsingForRequest::No;` like it was before you change. That should give you the behavior you want.

brentfulgham reviewed (changes requested):
You accidentally copy/pasted `return WebContentMode::Mobile`, when you just need to return what it was before (`return RecommendDesktopClassBrowsingForRequest::No;`)

brentfulgham commented on Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
```diff
@@ -1470,8 +1470,12 @@ static RecommendDesktopClassBrowsingForRequest desktopClassBrowsingRecommendedFo
+        policies.setCustomUserAgent(makeStringByReplacingAll(standardUserAgentWithApplicationName(m_applicationNameForUserAgent), "iPad"_s, "iPhone"_s));
+        policies.setCustomNavigatorPlatform("iPhone"_s);
         return RecommendDesktopClassBrowsingForRequest::No;
+    }
```
Perfect!

karlcow replied:
pfew. Thanks!

stwrt reviewed (approved):
Looks good

karlcow committed:
Adds UA Quirk for open.spotify.com on iPad Mini
https://bugs.webkit.org/show_bug.cgi?id=297710
rdar://158172517

Reviewed by Brandon Stewart and Brent Fulgham.

iPad Mini users are being blocked because of the UA string. This patch
makes it possible for iPad Mini users to receive the iPhone version of
the Spotify website, which would be better than the desktop version
given the form factor of the iPad Mini.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsIPhoneUserAgent):

Canonical link: https://commits.webkit.org/299301@main
